### PR TITLE
AP_RCProtocol: decode DSM frame markers to avoid timing related jitters

### PR DESCRIPTION
**Pull request by andyp1per**
This is pretty much a complete re-write using frame markers and betaflight's approach to decoding. It is very much more reliable for me than the format guess and seems to work well for others.

### CODE REVIEW

**Comment by WickedShell** [comment]
Doesn't this require a regeneration of the IOMCU firmwares as that runs the DSM decoder? (Not sure if you want to generate that, or if we should offload to @tridge since I think all the other ones that have been shipped were built on his environment)

**Comment by andyp1per** [comment]
Yes, and not sure this is entirely correct as yet. It should also check the frame markers in the sync state to go into potential de-sync

**Comment by Andrewiski** [comment]
@andyp1per I will try to do some testing this weekend on the beagle bone blue using dsm port with a spectrum receiver.

**Comment by Andrewiski** [comment]
I did a cherry pick on the three files but I am missing log_data method, assume I can just comment it out.

**Comment by andyp1per** [comment]
> 
> 
> I did a cherry pick on the three files but I am missing log_data method, assume I can just comment it out.

Yes

**Comment by Andrewiski** [comment]
So far has worked perfectly,  Setup is a Beagle Bone Blue, Spektrum SPM 4648 receiver and Spektrum DX8e transmitter. I cheyry picked the 3 files above, commented the log_data call and compiled.  All of my receiver issues went away. I tested in DSMX 11ms, dsm2 11ms, and dsm2 22ms at least according to the transmitter and all three worked.   Thanks for the effort and finding a fix

**Comment by andyp1per** [comment]
Flew with this this week - works fine

**Comment by tridge** [comment]
needs IO firmware re-gen

**Comment by andyp1per** [comment]
Ok I have rebased and regenerated the IOMCU firmware. @WickedShell are you able to test?

**Comment by WickedShell** [comment]
No problems noticed on a bench test with the following setup:
```
CubeBlack
DX8 G2
SPM9645
```

I bound and used DSM-X 11ms, and DSM-X 22ms without noting any problems.

**Comment by andyp1per** [comment]
Rebased and fixed RCProtocol tests - the Spektrum defined range is slightly different to what we have but the midpoint is the same.

File: libraries/AP_RCProtocol/AP_RCProtocol_DSM.cpp [hunk 1]
```diff
@@ -17,7 +17,12 @@
 /*
   with thanks to PX4 dsm.c for DSM decoding approach
```
Comment (tridge) [inline libraries/AP_RCProtocol/AP_RCProtocol_DSM.cpp:18]: this no longer has any PX4 code with your change, so we can remove this comment now

**Comment by tridge** [comment]
nice rewrite! good to go with minor comment change

**Pull request merged**